v0.45 Changelog: Improvements - Decompositions (#9418)
PennyLaneAI/pennylane: doc/releases/changelog-0.45.0.md

<h4>Decompositions 🍏</h4>

> * `qp.transforms.decompose` is now conveniently accessible from the top level as `qp.decompose`.

Reviewer: Could we make these sphinx references?

  [(#9011)](https://github.com/PennyLaneAI/pennylane/pull/9011)

* It is now possible to locally add decomposition rules to an operator via a `qp.decomposition.local_decomps` context
  manager. These rules will only take effect within the context.
  [(#8955)](https://github.com/PennyLaneAI/pennylane/pull/8955)
  [(#8998)](https://github.com/PennyLaneAI/pennylane/pull/8998)

  ```python
  @qp.register_resources({qp.CNOT: 1, qp.H: 2})
  def custom_decomp(wires):
      qp.H(wires[1])
      qp.CNOT(wires)
      qp.H(wires[1])
  ```
  ```pycon
  >>> with qp.decomposition.local_decomps():
  >>>    qp.add_decomps(qp.CZ, custom_decomp)
  >>>    print(qp.list_decomps(qp.CZ))
  Available Decomposition Rules:
  0: _cz_to_cps
  1: _cz_to_cnot
  2: _cz_to_ppr
  3: my_cz_rule
  ```
  ```
  >>> print(qp.list_decomps(qp.CZ))
  Available Decomposition Rules:
  0: _cz_to_cps
  1: _cz_to_cnot
  2: _cz_to_ppr
  ```

> * Several gates decompositions have been optimized for better performance:

Reviewer: ```suggestion
* Several decompositions have been optimized for resource efficiency:
```

>   - The `QROM` decompositions now has a smarter allocation of the work wires achieving better decompositions.

Reviewer: Can we use sphinx references for these as well?
```suggestion
  - The `QROM` decomposition now has a smarter allocation of the work wires achieving better decompositions.
```

    Note that this
    decomposition only is valid if `TemporaryAND` is used as intended - on zeroed input target qubits
    or zeroed output target qubits for `Adjoint(TemporaryAND)`.
    [(#9303)](https://github.com/PennyLaneAI/pennylane/pull/9303)

* A new decomposition of `Evolution` into `qp.PauliRot` has been added which is compatible with the
  new graph-based decomposition system. Similarly, a decomposition of `qp.RZ`
  into `qp.PhaseShift`, including a global phase, has been added.
  [(#9001)](https://github.com/PennyLaneAI/pennylane/pull/9001)
  [(#9049)](https://github.com/PennyLaneAI/pennylane/pull/9049)

* The graph-based decompositions system, enabled via :func:`~.decomposition.enable_graph`, now 
  additionally supports the custom `adjoint` method of qutrit operators such as ``QutritUnitary``,
  ``ControlledQutritUnitary``, and ``TRZ``.
  [(#9056)](https://github.com/PennyLaneAI/pennylane/pull/9056)

* The decomposition of :class:`~.QSVT` has been updated to be consistent with or without the graph-based
  decomposition system enabled.
  [(#8994)](https://github.com/PennyLaneAI/pennylane/pull/8994)

> * When the new graph-based decomposition system is enabled, the :func:`~pennylane.transforms.decompose`

Reviewer: These entries about changes to warnings can probably be moved to the bottom of the section.

Author: Good point, I've moved the 3 warning related entries to the bottom

  [(#9069)](https://github.com/PennyLaneAI/pennylane/pull/9069)
  [(#9124)](https://github.com/PennyLaneAI/pennylane/pull/9124)
  [(#9339)](https://github.com/PennyLaneAI/pennylane/pull/9339)

* The decompositions of :class:`~.TemporaryAND`, :class:`~.MultiRZ`, and :class:`~.DiagonalQubitUnitary`
  are now compatible with traced control values, traced wires, and traced data respectively
  [(#9157)](https://github.com/PennyLaneAI/pennylane/pull/9157)
  [(#9157)](https://github.com/PennyLaneAI/pennylane/pull/9157)
  [(#9157)](https://github.com/PennyLaneAI/pennylane/pull/9157)

* The :func:`~pennylane.ops.sk_decomposition` now accepts `"Adjoint(T)"` and `"Adjoint(S)"` in the `basis_set` as a
  now-preferred alternative to the old `"T*"` and `"S*"` convention for gate adjoints.
  [(#9231)](https://github.com/PennyLaneAI/pennylane/pull/9231)

* Some decomposition rules for ``MultiControlledX`` (e.g., ``_mcx_two_borrowed_workers`` and ``_mcx_one_borrowed_worker``)
  became identical for instances of this operator on less than 6 wires. To prevent this, stricter conditions have been applied on these decomposition.
  [(#9324)](https://github.com/PennyLaneAI/pennylane/pull/9324)

> * Removed some wire reusage in :class:`~.Select` that is not consistent with the approach to work

Reviewer: ```suggestion
* Removed some wire reusage in :class:`~.Select` that was not consistent with the approach to work
```

Author: I've applied this change here [1691da6](https://github.com/PennyLaneAI/pennylane/pull/9418/commits/1691da62103816023084aaa984df2cab441b6519)

<h4>Decompositions 🍏</h4>

* :func:`~pennylane.transforms.decompose` is now conveniently accessible from the top level as :func:`~pennylane.decompose`.

>   [(#9011)](https://github.com/PennyLaneAI/pennylane/pull/9011)

Author: ```suggestion
* :func:`pennylane.transforms.decompose` is now conveniently accessible from the top level as :func:`qp.decompose <pennylane.decompose>`.
[(#9011)](https://github.com/PennyLaneAI/pennylane/pull/9011)
```

  updated to use powers of ``X`` rather than ``RX``.
  [(#9069)](https://github.com/PennyLaneAI/pennylane/pull/9069)
  [(#9124)](https://github.com/PennyLaneAI/pennylane/pull/9124)
  [(#9339)](https://github.com/PennyLaneAI/pennylane/pull/9339)

* The decompositions of :class:`~.TemporaryAND`, :class:`~.MultiRZ`, and :class:`~.DiagonalQubitUnitary`
  are now compatible with Catalyst.
  [(#9157)](https://github.com/PennyLaneAI/pennylane/pull/9157)

* The :func:`~pennylane.ops.sk_decomposition` now accepts `"Adjoint(T)"` and `"Adjoint(S)"` in the `basis_set` as a
  now-preferred alternative to the old `"T*"` and `"S*"` convention for gate adjoints.
  [(#9231)](https://github.com/PennyLaneAI/pennylane/pull/9231)

* Some decomposition rules for :class:`~.MultiControlledX` (e.g., `_mcx_two_borrowed_workers` and
  `_mcx_one_borrowed_worker`) became identical for instances of this operator on less than 6 wires.
  To prevent this, stricter conditions have been applied on these decompositions.
  [(#9324)](https://github.com/PennyLaneAI/pennylane/pull/9324)

* Removed some wire reusage in :class:`~.Select` that is not consistent with the approach to work

>   wires elsewhere in PennyLane, and that was not taken into account in the resource functions

Author: ```suggestion
* Removed some wire reusage in :class:`~.Select` that was not consistent with the approach to work
wires elsewhere in PennyLane, and that was not taken into account in the resource functions
```

<h4>Decompositions 🍏</h4>

> * :func:`pennylane.transforms.decompose` is now conveniently accessible from the top level as :func:`qp.decompose <pennylane.decompose>`.

Reviewer 2: Very minor :) 

```suggestion
* :func:`qp.transforms.decompose <pennylane.transforms.decompose>` is now conveniently accessible from the top level as :func:`qp.decompose <pennylane.decompose>`.
```

[(#9011)](https://github.com/PennyLaneAI/pennylane/pull/9011)

> * It is now possible to locally add decomposition rules to an operator via a `qp.decomposition.local_decomps` context

Reviewer 2: ```suggestion
* It is now possible to locally add decomposition rules to an operator via a :func:`qp.decomposition.local_decomps <pennylane.decomposition.local_decomps>` context
```

>   manager. These rules will only take effect within the context.

Reviewer: ```suggestion
  manager. These rules will only be available within the context.
```

  - :class:`~.TemporaryAND` can now be decomposed into the equivalent (although slightly more expensive) :class:`~.Toffoli` gate.
    Note that this
    decomposition only is valid if ``TemporaryAND`` is used as intended - on zeroed input target qubits
    or zeroed output target qubits for ``Adjoint(TemporaryAND)``.
    [(#9303)](https://github.com/PennyLaneAI/pennylane/pull/9303)

  - A new decomposition of `Evolution` into `qp.PauliRot` has been added which is compatible with the
    new graph-based decomposition system. Similarly, a decomposition of `qp.RZ`
    into `qp.PhaseShift`, including a global phase, has been added.
    [(#9001)](https://github.com/PennyLaneAI/pennylane/pull/9001)
    [(#9049)](https://github.com/PennyLaneAI/pennylane/pull/9049)

* The graph-based decompositions system, enabled via :func:`~.decomposition.enable_graph`, now
  additionally supports the custom `adjoint` method of qutrit operators such as :class:`~.QutritUnitary`,
  :class:`~.ControlledQutritUnitary`, and :class:`~.TRZ`.
  [(#9056)](https://github.com/PennyLaneAI/pennylane/pull/9056)

* The decomposition of :class:`~.QSVT` has been updated to be consistent with or without the graph-based

>   decomposition system enabled.

Reviewer: ```suggestion
  decomposition system enabled (:func:`~.decomposition.enable_graph`).
```

  [(#8994)](https://github.com/PennyLaneAI/pennylane/pull/8994)

* Now, when the new graph-based decomposition system is enabled, the :func:`~pennylane.transforms.decompose`
  transform no longer tries to find a decomposition for an operator that meets a provided ``stopping_condition``,
  even if it is not in the defined ``gate_set``.
  [(#9036)](https://github.com/PennyLaneAI/pennylane/pull/9036)

> * Added a `strict` keyword to the :func:`~pennylane.transforms.decompose` transform that, when set to ``False``,

Reviewer 2: ```suggestion
* A `strict` keyword argument was added to the :func:`~pennylane.decompose` transform that, when set to ``False``,
```

also very minor :^)

  wires and/or states. Additionally, the parametric decomposition for traced states without `qjit` was
  updated to use powers of ``X`` rather than ``RX``.
  [(#9069)](https://github.com/PennyLaneAI/pennylane/pull/9069)
  [(#9124)](https://github.com/PennyLaneAI/pennylane/pull/9124)
  [(#9339)](https://github.com/PennyLaneAI/pennylane/pull/9339)

* The decompositions of :class:`~.TemporaryAND`, :class:`~.MultiRZ`, and :class:`~.DiagonalQubitUnitary`
  are now compatible with Catalyst.
  [(#9157)](https://github.com/PennyLaneAI/pennylane/pull/9157)

* The :func:`~pennylane.ops.sk_decomposition` now accepts `"Adjoint(T)"` and `"Adjoint(S)"` in the `basis_set` as a
  now-preferred alternative to the old `"T*"` and `"S*"` convention for gate adjoints.
  [(#9231)](https://github.com/PennyLaneAI/pennylane/pull/9231)

* Some decomposition rules for :class:`~.MultiControlledX` (e.g., `_mcx_two_borrowed_workers` and
  `_mcx_one_borrowed_worker`) became identical for instances of this operator on less than 6 wires.
  To prevent this, stricter conditions have been applied on these decompositions.
  [(#9324)](https://github.com/PennyLaneAI/pennylane/pull/9324)

> * Removed some wire reusage in :class:`~.Select` that was not consistent with the approach to work

Reviewer 2: ```suggestion
* Some wire reusage was removed in :class:`~.Select` that was not consistent with the approach to work
```
